subiquity/server/controllers/cmdlist.py

```diff
@@ -124,7 +127,7 @@ def __init__(self, app):
                 ]
 
     def env(self):
-        env = super().env()
+        env = orig_environ(None)
```

Comment:
Our [docs](https://github.com/canonical/subiquity/blob/9bf0a50a24b08ba7f80277212acf10a380ef03b0/documentation/autoinstall-reference.md#late-commands) indicate:

> You can run curtin in-target -- $shell_command [...] to run in the target system

and I think sadly your change will break the ability to do that. Maybe we need to update the docs, or add support for specifying if we want a clean env or not, as in:

```yaml
late-commands:
    # specifies a command directly (runs with a clean env)
    - echo "hello world"
    # Run `cat /etc/machine-id` in the target system
    - cmd: curtin in-target -- cat /etc/machine-id
      clean-env: false  # defaults to true? (true feels more natural but false might be better for backward compatibility)
``` 

or maybe we add the ability to run a command in the target system:

```yaml
late-commands:
    # specifies a command directly (runs in the installer environment)
    - echo "hello world"
    # Run `cat /etc/machine-id` in the target system
    - cmd: cat /etc/machine-id
      in-target: true    # defaults to false
```

in any case, I think we need to think about what it means in terms of backward compatibility

Reply:
Another option (which is not my favorite!) would be to check if the command starts with "curtin":

```python
if shlex.split(cmd.desc())[0] == "curtin":
    clean_env = False
```

Reply:
I do think `shlex.split()` is where this needs to go.  Maybe we also pair that with `shutil.which()` and check if that resolves to something in the snap.

Reply:
Thanks for pointing this out. For the general case, I like the idea of inspecting the path from `shutil.which()` and determining if we should drop the snap environment variables or not.

Eventually adding support in the autoinstall config to specify the command environment seems like a good call, although perhaps we could encourage the use of `env` until we decide if want to and how that would look.